clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/AbstractPubSubBasedMessageSender.java

```diff
@@ -0,0 +1,306 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.PubsubMessage;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.tag.Tags;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.EncodeException;
+import io.vertx.core.json.Json;
+import io.vertx.ext.healthchecks.HealthCheckHandler;
+import io.vertx.ext.healthchecks.Status;
+
+/**
+ * A sender for publishing messages to Pub/Sub.
+ */
+public abstract class AbstractPubSubBasedMessageSender implements MessagingClient, ServiceClient, Lifecycle {
+
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+    private final PubSubPublisherFactory publisherFactory;
+    private final String topic;
+    private final Tracer tracer;
+    private final String projectId;
+    private LifecycleStatus lifecycleStatus = new LifecycleStatus();
+
+    /**
+     * Creates a new PubSub-based message sender.
+     *
+     * @param publisherFactory The factory to use for creating Pub/Sub publishers.
+     * @param topic The topic to create the publisher for.
+     * @param projectId The Google project id to use.
+     * @param tracer The OpenTracing tracer.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public AbstractPubSubBasedMessageSender(
+            final PubSubPublisherFactory publisherFactory,
+            final String topic,
+            final String projectId,
+            final Tracer tracer) {
+        Objects.requireNonNull(publisherFactory);
+        Objects.requireNonNull(topic);
+        Objects.requireNonNull(projectId);
+        Objects.requireNonNull(tracer);
+
+        this.publisherFactory = publisherFactory;
+        this.topic = topic;
+        this.projectId = projectId;
+        this.tracer = tracer;
+    }
+
+    /**
+     * Creates a new PubSub-based message sender.
+     * <p>
+     * To be used for unit tests.
+     *
+     * @param publisherFactory The factory to use for creating Pub/Sub publishers.
+     * @param topic The topic to create the publisher for.
+     * @param projectId The Google project id to use.
+     * @param tracer The OpenTracing tracer.
+     * @param lifecycleStatus The lifecycle to use for this class.
+     */
+    protected AbstractPubSubBasedMessageSender(
+            final PubSubPublisherFactory publisherFactory,
+            final String topic,
+            final String projectId,
+            final Tracer tracer,
+            final LifecycleStatus lifecycleStatus) {
+
+        this.publisherFactory = publisherFactory;
+        this.topic = topic;
+        this.projectId = projectId;
+        this.tracer = tracer;
+        this.lifecycleStatus = lifecycleStatus;
+    }
+
+    private static Span newSpan(final Tracer tracer, final String operationName, final String topic,
```

Comment:
Are the passed in tracer and topic name expected to differ from the values stored in the class' fields?

Reply:
No they are not. Removed passed parameters and use class parameters